pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/intervalST/IntervalST.java

```diff
@@ -0,0 +1,187 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.broker.routing.segmentpruner.intervalST;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+
+// A read-only interval search tree based on AVL Tree
+public class IntervalST<Value> {
+
+  private Node<Value> root;
+
+  public IntervalST(List<Long> mins, List<Long> maxs, List<Value> values) {
```

Comment:
Since this is a read-only interval search tree, we can simplify the tree construction with the following steps:
1. Sort the intervals
2. Construct a full BST from the sorted intervals
3. Use DFS to generate the auxiliary info

One step further, you can skip the step 2 and directly use the sorted intervals as the tree structure and use index to access the full BST (calculate the index is a little bit tricky though)

Reply:
Cool idea! I think we can have a list representation of the bst (like array representation of heap). The root will be indexed as 0, left child 1, right child 2 and so on. The root of each subtree will be the one has median interval for that subtree. 
```
A typical balanced tree
                              [10, 20]
                              /       \
                       [8, 15]        [12, 20]  
                          /            / 
                   [5, 10]       [10, 30]

will be represented as { [10, 20], [8, 15], [12, 20], [5, 10], null, [10, 30] }
```